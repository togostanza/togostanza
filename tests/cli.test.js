import os from 'os';
import path from 'path';
import { spawnSync } from 'child_process';

import fixturify from 'fixturify';
import fs from 'fs-extra';
import walkSync from 'walk-sync';

test('--help', () => {
  const {stdout, status} = togostanza(['--help']);

  expect(stdout).toContain('Usage: togostanza [options] [command]');
  expect(status).toBe(0);
});

describe('init', () => {
  test.each([
    {packageManager: 'npm'},
    {packageManager: 'yarn'}
  ])('non-interactive (%p)', ({packageManager}) => {
    withinTmpdir(() => {
      const {output, status} = togostanza(['init',
        '--git-url',         'https://github.com/ursm/some-repo.git',
        '--name',            'some-repo',
        '--license',         'MIT',
        '--package-manager', packageManager,
        '--skip-install',
        '--skip-git'
      ]);

      expect(output).toMatchSnapshot();
      expect(status).toBe(0);

      expect(fixturify.readSync('.')).toMatchSnapshot();
    });
  });

  test('with scp-like repository url', () => {
    withinTmpdir(() => {
      const {output, status} = togostanza(['init',
        '--git-url',         'git@github.com:ursm/some-repo.git',
        '--name',            'some-repo',
        '--license',         'MIT',
        '--package-manager', 'npm',
        '--skip-install',
        '--skip-git'
      ]);

      expect(output).toMatchSnapshot();
      expect(status).toBe(0);

      expect(fixturify.readSync('.')).toMatchSnapshot();
    });
  });
});

describe('generate stanza', () => {
  test('non-interactive', () => {
    withinTmpdir(() => {
      const {output, status} = togostanza(['generate', 'stanza', 'hello',
        '--label',      'LABEL',
        '--definition', 'DEFINITION',
        '--type',       'TYPE',
        '--provider',   'PROVIDER',
        '--license',    'LICENSE',
        '--author',     'AUTHOR',
        '--address',    'ADDRESS',
        '--timestamp',  '2020-10-05'
      ]);

      expect(output).toMatchSnapshot();
      expect(status).toBe(0);

      expect(fixturify.readSync('.')).toMatchSnapshot();
    });
  });

  test('with camelCaseId', () => {
    withinTmpdir(() => {
      const {output, status} = togostanza(['generate', 'stanza', 'helloWorld',
        '--label',      'LABEL',
        '--definition', 'DEFINITION',
        '--type',       'TYPE',
        '--provider',   'PROVIDER',
        '--license',    'LICENSE',
        '--author',     'AUTHOR',
        '--address',    'ADDRESS',
        '--timestamp',  '2020-10-05'
      ]);

      expect(output).toMatchSnapshot();
      expect(status).toBe(0);

      expect(fixturify.readSync('.')).toMatchSnapshot();
    });
  });
});

describe('upgrade', () => {
  test('move stanza into stanzas/', () => {
    withinTmpdir(() => {
      fixturify.writeSync('.', {
        hello: {
          'metadata.json': '{}'
        }
      });

      const {stdout, stderr, status} = togostanza(['upgrade']);

      expect(stdout.replace(/^Time elapsed: .*$/m, '')).toMatchSnapshot();
      expect(stderr).toMatchSnapshot();
      expect(status).toBe(0);

      expect(fixturify.readSync('.')).toMatchSnapshot();
    });
  });

  test('convert Stanza() to default exported function', () => {
    withinTmpdir(() => {
      fixturify.writeSync('stanzas/hello', {
        'index.js': 'Stanza(function(stanza, params) {});'
      })

      const {stdout, status} = togostanza(['upgrade']);

      expect(stdout).toContain('1 ok');
      expect(status).toBe(0);

      expect(fs.readFileSync('stanzas/hello/index.js', 'utf8')).toBe('export default function hello(stanza, params) {}')
    });
  });
});

describe('build', () => {
  test('simple', () => {
    withinTmpdir(() => {
      const init = togostanza(['init',
        '--git-url',         '',
        '--name',            'some-repo',
        '--license',         'MIT',
        '--package-manager', 'yarn',
        '--skip-install',
        '--skip-git',
      ]);

      expect(init.stderr).toContain('Getting Started');
      expect(init.status).toBe(0);

      process.chdir('some-repo');

      fixturify.writeSync('.', {
        assets: {
          foo: 'FOO'
        }
      });

      const generateStanza = togostanza(['generate', 'stanza', 'hello',
        '--label',      'LABEL',
        '--definition', 'DEFINITION',
        '--type',       'TYPE',
        '--provider',   'PROVIDER',
        '--license',    'LICENSE',
        '--author',     'AUTHOR',
        '--address',    'ADDRESS',
        '--timestamp',  '2020-10-05'
      ]);

      expect(generateStanza.output).toMatchSnapshot();
      expect(generateStanza.status).toBe(0);

      fixturify.writeSync('stanzas/hello', {
        assets: {
          bar: 'BAR'
        }
      });

      const build = togostanza(['build'], {timeout: 30_000});

      expect(build.status).toBe(0);

      expect(walkSync('dist', {
        ignore: [
          'node_modules'
        ]
      }).map(fn => {
        return fn.replace(/^(-togostanza\/Layout-)[0-9a-f]+(\.js(?:\.map)?)$/, '$1HASH$2');
      })).toMatchSnapshot();

      expect(fs.readFileSync('dist/hello.js', 'utf8')).toContain("import Stanza from 'togostanza/stanza';");
    });
  });
});

function togostanza(args, opts = {}) {
  const bin = path.resolve(__dirname, '../bin/togostanza.mjs');

  // --no-warnings is necessary to avoid including ESM warnings in the snapshot, which change with each test run
  // (not required in Node.js 14)
  const ps = spawnSync(process.argv0, ['--no-warnings', bin, ...args], {
    encoding: 'utf8',
    timeout: 30_000,
    env: {
      ...process.env,
      FORCE_COLOR: '0'
    },
    ...opts
  });

  expect(ps.error).toBeUndefined();

  return ps;
}

function withinTmpdir(cb) {
  const tmp = fs.mkdtempSync(path.join(os.tmpdir(), 'togostanza-'));

  try {
    const cwd = process.cwd();

    process.chdir(tmp);

    try {
      cb(tmp);
    } finally {
      process.chdir(cwd);
    }
  } finally {
    fs.removeSync(tmp);
  }
}
