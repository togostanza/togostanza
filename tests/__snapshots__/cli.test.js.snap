// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build simple 1`] = `
Array [
  null,
  "",
  "   create stanzas/hello/metadata.json
   create stanzas/hello/README.md
   create stanzas/hello/index.js
   create stanzas/hello/style.scss
   create stanzas/hello/assets/.keep
   create stanzas/hello/templates/stanza.html.hbs
",
]
`;

exports[`build simple 2`] = `
Array [
  "-togostanza/",
  "-togostanza/Layout-HASH.js",
  "-togostanza/Layout-HASH.js.map",
  "-togostanza/help-app.js",
  "-togostanza/help-app.js.map",
  "-togostanza/index-app.js",
  "-togostanza/index-app.js.map",
  "assets/",
  "assets/foo",
  "hello.css",
  "hello.html",
  "hello.js",
  "hello.js.map",
  "hello/",
  "hello/assets/",
  "hello/assets/bar",
  "hello/metadata.json",
  "index.html",
]
`;

exports[`generate stanza non-interactive 1`] = `
Array [
  null,
  "",
  "   create stanzas/hello/metadata.json
   create stanzas/hello/README.md
   create stanzas/hello/index.js
   create stanzas/hello/style.scss
   create stanzas/hello/assets/.keep
   create stanzas/hello/templates/stanza.html.hbs
",
]
`;

exports[`generate stanza non-interactive 2`] = `
Object {
  "stanzas": Object {
    "hello": Object {
      "README.md": "# LABEL

Stanza description goes here. Edit \`stanzas/hello/README.md\` to update.
",
      "assets": Object {
        ".keep": "",
      },
      "index.js": "import Stanza from 'togostanza/stanza';

export default class Hello extends Stanza {
  async render() {
    this.renderTemplate(
      {
        template: 'stanza.html.hbs',
        parameters: {
          greeting: \`Hello, \${this.params['say-to']}!\`
        }
      }
    );
  }
}
",
      "metadata.json": "{
  \\"@context\\": {
    \\"stanza\\": \\"http://togostanza.org/resource/stanza#\\"
  },
  \\"@id\\": \\"hello\\",
  \\"stanza:label\\": \\"LABEL\\",
  \\"stanza:definition\\": \\"DEFINITION\\",
  \\"stanza:license\\": \\"LICENSE\\",
  \\"stanza:author\\": \\"AUTHOR\\",
  \\"stanza:contributor\\": [],
  \\"stanza:created\\": \\"2020-10-05\\",
  \\"stanza:updated\\": \\"2020-10-05\\",
  \\"stanza:parameter\\": [
    {
      \\"stanza:key\\": \\"say-to\\",
      \\"stanza:type\\": \\"string\\",
      \\"stanza:example\\": \\"world\\",
      \\"stanza:description\\": \\"who to say hello to\\",
      \\"stanza:required\\": false
    }
  ],
  \\"stanza:menu-placement\\": \\"bottom-right\\",
  \\"stanza:style\\": [
    {
      \\"stanza:key\\": \\"--greeting-color\\",
      \\"stanza:type\\": \\"color\\",
      \\"stanza:default\\": \\"#eb7900\\",
      \\"stanza:description\\": \\"text color of greeting\\"
    },
    {
      \\"stanza:key\\": \\"--greeting-align\\",
      \\"stanza:type\\": \\"single-choice\\",
      \\"stanza:choice\\": [
        \\"left\\",
        \\"center\\",
        \\"right\\"
      ],
      \\"stanza:default\\": \\"center\\",
      \\"stanza:description\\": \\"text align of greeting\\"
    }
  ],
  \\"stanza:incomingEvent\\": [],
  \\"stanza:outgoingEvent\\": []
}
",
      "style.scss": "// Load the repository-wide global style here. The actual file is located at ../common.scss for details.
@use '@/common.scss';

// The following is just an example. Feel free to modify it.
// --greeting-color and --greeting-align are style variables, which are defined in metadata.json.
main {
  padding: 1rem 2rem;
}

p.greeting {
  margin: 0;
  font-size: 24px;
  color: var(--greeting-color);
  text-align: var(--greeting-align);
}
",
      "templates": Object {
        "stanza.html.hbs": "<p class=\\"greeting\\">{{greeting}}</p>
",
      },
    },
  },
}
`;

exports[`generate stanza with camelCaseId 1`] = `
Array [
  null,
  "",
  "   create stanzas/hello-world/metadata.json
   create stanzas/hello-world/README.md
   create stanzas/hello-world/index.js
   create stanzas/hello-world/style.scss
   create stanzas/hello-world/assets/.keep
   create stanzas/hello-world/templates/stanza.html.hbs
",
]
`;

exports[`generate stanza with camelCaseId 2`] = `
Object {
  "stanzas": Object {
    "hello-world": Object {
      "README.md": "# LABEL

Stanza description goes here. Edit \`stanzas/hello-world/README.md\` to update.
",
      "assets": Object {
        ".keep": "",
      },
      "index.js": "import Stanza from 'togostanza/stanza';

export default class HelloWorld extends Stanza {
  async render() {
    this.renderTemplate(
      {
        template: 'stanza.html.hbs',
        parameters: {
          greeting: \`Hello, \${this.params['say-to']}!\`
        }
      }
    );
  }
}
",
      "metadata.json": "{
  \\"@context\\": {
    \\"stanza\\": \\"http://togostanza.org/resource/stanza#\\"
  },
  \\"@id\\": \\"hello-world\\",
  \\"stanza:label\\": \\"LABEL\\",
  \\"stanza:definition\\": \\"DEFINITION\\",
  \\"stanza:license\\": \\"LICENSE\\",
  \\"stanza:author\\": \\"AUTHOR\\",
  \\"stanza:contributor\\": [],
  \\"stanza:created\\": \\"2020-10-05\\",
  \\"stanza:updated\\": \\"2020-10-05\\",
  \\"stanza:parameter\\": [
    {
      \\"stanza:key\\": \\"say-to\\",
      \\"stanza:type\\": \\"string\\",
      \\"stanza:example\\": \\"world\\",
      \\"stanza:description\\": \\"who to say hello to\\",
      \\"stanza:required\\": false
    }
  ],
  \\"stanza:menu-placement\\": \\"bottom-right\\",
  \\"stanza:style\\": [
    {
      \\"stanza:key\\": \\"--greeting-color\\",
      \\"stanza:type\\": \\"color\\",
      \\"stanza:default\\": \\"#eb7900\\",
      \\"stanza:description\\": \\"text color of greeting\\"
    },
    {
      \\"stanza:key\\": \\"--greeting-align\\",
      \\"stanza:type\\": \\"single-choice\\",
      \\"stanza:choice\\": [
        \\"left\\",
        \\"center\\",
        \\"right\\"
      ],
      \\"stanza:default\\": \\"center\\",
      \\"stanza:description\\": \\"text align of greeting\\"
    }
  ],
  \\"stanza:incomingEvent\\": [],
  \\"stanza:outgoingEvent\\": []
}
",
      "style.scss": "// Load the repository-wide global style here. The actual file is located at ../common.scss for details.
@use '@/common.scss';

// The following is just an example. Feel free to modify it.
// --greeting-color and --greeting-align are style variables, which are defined in metadata.json.
main {
  padding: 1rem 2rem;
}

p.greeting {
  margin: 0;
  font-size: 24px;
  color: var(--greeting-color);
  text-align: var(--greeting-align);
}
",
      "templates": Object {
        "stanza.html.hbs": "<p class=\\"greeting\\">{{greeting}}</p>
",
      },
    },
  },
}
`;

exports[`init non-interactive ({"packageManager": "npm"}) 1`] = `
Array [
  null,
  "",
  "   create package.json
   create README.md
   create .gitignore
   create common.scss
   create assets/.keep
   create lib/.keep
   create .github/workflows/publish.yml
Skipping install command: npm install --cache-min 86400


I'm all done. Just run npm install to install the required dependencies.



Getting Started
---------------

Create a new stanza:

  $ cd some-repo
  $ npx togostanza generate stanza

Serve the repository locally:

  $ cd some-repo
  $ npx togostanza serve

Build stanzas for deployment:

  $ cd some-repo
  $ npx togostanza build

",
]
`;

exports[`init non-interactive ({"packageManager": "npm"}) 2`] = `
Object {
  "some-repo": Object {
    ".github": Object {
      "workflows": Object {
        "publish.yml": "name: Publish GitHub Pages

on:
  push:
    branches:
      - main

jobs:
  deploy:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2
        with:
          submodules: true
          fetch-depth: 0

      - uses: actions/setup-node@v1
        with:
          node-version: '14'

      - run: npm ci
      - run: npx togostanza build

      - uses: peaceiris/actions-gh-pages@v3
        with:
          github_token: \${{ secrets.GITHUB_TOKEN }}
          publish_dir: ./dist
",
      },
    },
    ".gitignore": "/dist/
/node_modules/
",
    "README.md": "# some-repo

Visit https://ursm.github.io/some-repo/ to see the stanzas in this repository in action.

This was generated by [TogoStanza](http://togostanza.org), which enables the development of reusable web components that are embeddable into any web pages.

See [togostanza's README.md](https://github.com/togostanza/togostanza#readme) for details.

## To repository author

Describe about the repository here. These contents will help the repository users:

- What is this repository for?
- Who maintains this repository?
- ...
",
    "assets": Object {
      ".keep": "",
    },
    "common.scss": "/*

You can set up a global style here that is commonly used in each stanza.

Example:

h1 {
  font-size: 24px;
}

*/
",
    "lib": Object {
      ".keep": "",
    },
    "package.json": "{
  \\"name\\": \\"some-repo\\",
  \\"version\\": \\"0.0.1\\",
  \\"license\\": \\"MIT\\",
  \\"repository\\": \\"https://github.com/ursm/some-repo.git\\",
  \\"dependencies\\": {
    \\"togostanza\\": \\"^3.0.0-beta.38\\"
  },
  \\"engines\\": {
    \\"node\\": \\">=14\\"
  },
  \\"private\\": true
}
",
  },
}
`;

exports[`init non-interactive ({"packageManager": "yarn"}) 1`] = `
Array [
  null,
  "",
  "   create package.json
   create README.md
   create .gitignore
   create common.scss
   create assets/.keep
   create lib/.keep
   create .github/workflows/publish.yml
Skipping install command: yarn install


I'm all done. Just run yarn install to install the required dependencies.



Getting Started
---------------

Create a new stanza:

  $ cd some-repo
  $ yarn run togostanza generate stanza

Serve the repository locally:

  $ cd some-repo
  $ yarn run togostanza serve

Build stanzas for deployment:

  $ cd some-repo
  $ yarn run togostanza build

",
]
`;

exports[`init non-interactive ({"packageManager": "yarn"}) 2`] = `
Object {
  "some-repo": Object {
    ".github": Object {
      "workflows": Object {
        "publish.yml": "name: Publish GitHub Pages

on:
  push:
    branches:
      - main

jobs:
  deploy:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2
        with:
          submodules: true
          fetch-depth: 0

      - uses: actions/setup-node@v1
        with:
          node-version: '14'

      - run: yarn install --frozen-lockfile
      - run: yarn run togostanza build

      - uses: peaceiris/actions-gh-pages@v3
        with:
          github_token: \${{ secrets.GITHUB_TOKEN }}
          publish_dir: ./dist
",
      },
    },
    ".gitignore": "/dist/
/node_modules/
",
    "README.md": "# some-repo

Visit https://ursm.github.io/some-repo/ to see the stanzas in this repository in action.

This was generated by [TogoStanza](http://togostanza.org), which enables the development of reusable web components that are embeddable into any web pages.

See [togostanza's README.md](https://github.com/togostanza/togostanza#readme) for details.

## To repository author

Describe about the repository here. These contents will help the repository users:

- What is this repository for?
- Who maintains this repository?
- ...
",
    "assets": Object {
      ".keep": "",
    },
    "common.scss": "/*

You can set up a global style here that is commonly used in each stanza.

Example:

h1 {
  font-size: 24px;
}

*/
",
    "lib": Object {
      ".keep": "",
    },
    "package.json": "{
  \\"name\\": \\"some-repo\\",
  \\"version\\": \\"0.0.1\\",
  \\"license\\": \\"MIT\\",
  \\"repository\\": \\"https://github.com/ursm/some-repo.git\\",
  \\"dependencies\\": {
    \\"togostanza\\": \\"^3.0.0-beta.38\\"
  },
  \\"engines\\": {
    \\"node\\": \\">=14\\"
  },
  \\"private\\": true
}
",
  },
}
`;

exports[`init with scp-like repository url 1`] = `
Array [
  null,
  "",
  "   create package.json
   create README.md
   create .gitignore
   create common.scss
   create assets/.keep
   create lib/.keep
   create .github/workflows/publish.yml
Skipping install command: npm install --cache-min 86400


I'm all done. Just run npm install to install the required dependencies.



Getting Started
---------------

Create a new stanza:

  $ cd some-repo
  $ npx togostanza generate stanza

Serve the repository locally:

  $ cd some-repo
  $ npx togostanza serve

Build stanzas for deployment:

  $ cd some-repo
  $ npx togostanza build

",
]
`;

exports[`init with scp-like repository url 2`] = `
Object {
  "some-repo": Object {
    ".github": Object {
      "workflows": Object {
        "publish.yml": "name: Publish GitHub Pages

on:
  push:
    branches:
      - main

jobs:
  deploy:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2
        with:
          submodules: true
          fetch-depth: 0

      - uses: actions/setup-node@v1
        with:
          node-version: '14'

      - run: npm ci
      - run: npx togostanza build

      - uses: peaceiris/actions-gh-pages@v3
        with:
          github_token: \${{ secrets.GITHUB_TOKEN }}
          publish_dir: ./dist
",
      },
    },
    ".gitignore": "/dist/
/node_modules/
",
    "README.md": "# some-repo

Visit https://ursm.github.io/some-repo/ to see the stanzas in this repository in action.

This was generated by [TogoStanza](http://togostanza.org), which enables the development of reusable web components that are embeddable into any web pages.

See [togostanza's README.md](https://github.com/togostanza/togostanza#readme) for details.

## To repository author

Describe about the repository here. These contents will help the repository users:

- What is this repository for?
- Who maintains this repository?
- ...
",
    "assets": Object {
      ".keep": "",
    },
    "common.scss": "/*

You can set up a global style here that is commonly used in each stanza.

Example:

h1 {
  font-size: 24px;
}

*/
",
    "lib": Object {
      ".keep": "",
    },
    "package.json": "{
  \\"name\\": \\"some-repo\\",
  \\"version\\": \\"0.0.1\\",
  \\"license\\": \\"MIT\\",
  \\"repository\\": \\"https://github.com/ursm/some-repo.git\\",
  \\"dependencies\\": {
    \\"togostanza\\": \\"^3.0.0-beta.38\\"
  },
  \\"engines\\": {
    \\"node\\": \\">=14\\"
  },
  \\"private\\": true
}
",
  },
}
`;

exports[`upgrade move stanza into stanzas/ 1`] = `
"move  hello -> stanzas/hello

No files selected, nothing to do. 
All done. 
Results: 
0 errors
0 unmodified
0 skipped
0 ok

"
`;

exports[`upgrade move stanza into stanzas/ 2`] = `""`;

exports[`upgrade move stanza into stanzas/ 3`] = `
Object {
  "stanzas": Object {
    "hello": Object {
      "metadata.json": "{}",
    },
  },
}
`;
